keps/sig-node/5067-pod-generation/README.md

```diff
@@ -0,0 +1,971 @@
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+Because there is only one singular `status.observedGeneration` in the pod status, only
+one writer can set it. This is consistent with other object types that have
+`status.observedGeneration`, and the expectation is that the primary controller for
+the object sets the field. For a pod, that primary controller is the kubelet, so
+once a pod is bound to a node, we expect that the kubelet on that node is the
+sole writer of `status.observedGeneration`. 
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+Today, it is possible for a client to custom-set a pod's `metadata.generation`, but once
+set, the `metadata.generation` cannot be updated. That means that there is a theoretical 
+possibility for an external client to be setting `metadata.generation` on pod 
+create and depend on that fixed value somehow in its own reconciliation logic.
+
+That said, the `metadata.generation` field is described as ["set by the system and monotonically increasing"](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata),
+and thus it should be clear enough that `metadata.generation` was not intended to be
+used in this way.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+#### API server
+
+For a newly created pod, the API server will set `metadata.generation` to 1. For any updates
+to the [PodSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#podspec-v1-core),
+and additionally updates to annotations that end up getting copied to the PodSpec
+(such as seccomp and AppArmor annotations), the API server will increment 
+`metadata.generation` by 1.
+
+The pod `metadata.generation` will also continue to be incremented on graceful delete or 
+deferred delete, just as the API server currently does for pods and other 
+objects today.
+
+Pod updates that would not result in `metadata.generation` being incremented include:
+* Changes to metadata. This means that if a Pod uses the downward API to make
+pod metadata available to containers, Pod behavior can change without the
+generation being incremented. We will consider this working as intended.
+* Changes to status.
+
+Any attempts by clients to set or modify the `metadata.generation` field themselves will be
```

Comment:
will we allow resetting it to empty? So older webhooks that strip out this field will not break the status update?

Reply:
This field has existed for a long time, and IIUC is immutable from the client, so I don't think this is an issue.

Reply:
but if existing webhook sets it to empty expliticly, it will work today, but will break once we enforce this check. I think it may be a real issue, I just don't know how prominent it will be. Maybe some api reviewer can help here

Reply:
IIUC, the field will be overwritten on creation (so any value ignored). On update, mutation was never allowed, so it should be a non-issue. I don't think there's anything new being enforced here (@natasha41575 correct me if I'm wrong)

>  if existing webhook sets it to empty expliticly, it will work today

There are edge cases that can set it to a non-empty value today, so this could already run in to trouble.

Reply:
if we add a validation that doesn't allow to reset to empty, than existing mutator that was setting it to nil explicitly (with no effect in the past) will start failing as now it will be resetting the value.

Reply:
When updating status, if the incoming update clears observedGeneration, we should probably just preserve the existing value. We've done that for other pod fields IIRC.

Reply:
The apiserver will ignore whatever the client has set `metadata.generation` to, and overwrite it with the correct value. That includes if clients strip it out - the apiserver will just reset it to whatever it was before the update request + 1. So this shouldn't break older webhooks.

Reply:
Just tried it out experimentally again to confirm - today, if I set pod `metadata.generation` to some nonzero value on `pod create`, and I attempt to update it (either by explicitly setting it to `nil` or setting it to some other value`), my attempt to reset does not result in failure, it is just ignored and I don't see the generation being updated.

The proposal here is to keep that same "ignore" behavior, so older webhooks should not be impacted.

Reply:
> Maybe some api reviewer can help here

@liggitt do you want to chime in here?

Reply:
> but if existing webhook sets it to empty expliticly, it will work today, but will break once we enforce this check. I think it may be a real issue, I just don't know how prominent it will be. Maybe some api reviewer can help here

Are you referring to the `metadata.generation` field?

If a webhook is setting it on create, today that is allowed to persist. With the changes in this KEP, will now be overwritten with a `1`, but will not fail the request.

If a webhook is setting that on update, today that is ignored and the existing value is kept silently. With the changes in this KEP, that behavior is preserved, unless a spec field is changed, in which case the field will be incremented by 1. In neither case does this fail the request.

Reply:
cool. as long as we are overwriting it, and not failing on bad values, this sounds good